ucp/useraccounts_module.php

```diff
@@ -319,57 +323,97 @@ function main($id, $mode)
+					$delete_sql = empty($markedbadges) ? '' : ' AND ' . $this->db->sql_in_set('badge_id', $markedbadges, true); // true for NOT IN, false (default) for IN
 					$sql = 'DELETE FROM ' . $user_badge_table . '
 						WHERE user_id = ' . $this_user_id . "
 						$delete_sql";
-					$this->db->sql_query($sql);
+					if (!$this->db->sql_query($sql))
+					{
+						trigger_error($this->user->lang('UCP_DELETE_UNSELECTED_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+					}
 
-					foreach ($markedbadges as $markedbadge)
+					if ($markedbadges)
 					{
-						// Reject duplicates.
-						$sql = "SELECT COUNT(*) as duplcount
-							FROM $user_badge_table
-							WHERE badge_id = $markedbadge
-							AND user_id = $this_user_id";
+						// Query valid badges
+						$sql = "SELECT badge_id
+							FROM $badge_table
+							WHERE badge_default = TRUE";
 						if (!($result = $this->db->sql_query($sql)))
 						{
-							trigger_error($this->user->lang('UCP_DUPLICATE_USER_BADGE_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+							trigger_error($this->user->lang('UCP_DEFAULT_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
 						}
-						$duplicates = (int) $this->db->sql_fetchfield('duplcount');
+						$default_badges = $this->db->sql_fetchrowset($result);
 						$this->db->sql_freeresult($result);
-						if (!$duplicates)
+
+						$sql = "SELECT badge_id
+							FROM $badge_avail_table
+							WHERE user_id = $this_user_id";
+						if (!($result = $this->db->sql_query($sql)))
 						{
-							// Add a new user badge.
-							$data = array(
-								'user_id' => $this_user_id,
-								'badge_id' => $markedbadge,
-								'badge_order' => null // Order value will be validated after loop.
-							);
-							$sql = 'INSERT INTO ' . $user_badge_table . $this->db->sql_build_array('INSERT', $data);
-							$this->db->sql_query($sql);
+							trigger_error($this->user->lang('UCP_AVAIL_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
 						}
+						$custom_badges = $this->db->sql_fetchrowset($result);
+						$this->db->sql_freeresult($result);
+
+						// Prepare an array for inserting badges
+						$sql_array = array();
+						foreach ($markedbadges as $markedbadge)
+						{
+							// Validate requested badges and reject duplicates.
+							if (($this->core->in_2d_array($markedbadge, $default_badges) || $this->core->in_2d_array($markedbadge, $custom_badges)) && !$this->core->in_2d_array($markedbadge, $user_badges))
+							{
+								$sql_array[] = array(
+									'user_id' => $this_user_id,
+									'badge_id' => $markedbadge,
+									'badge_order' => null // Order value will be validated after loop.
+								);
+							}
+						}
+
+						if(!empty($sql_array))
```

Comment:
Style nit: need a space between `if (`

Reply:
On closer inspection `$this->db->sql_multi_insert` already includes a check for empty input, so this whole block can simplify down to `$this->db->sql_multi_insert($user_badge_table, $sql_array);` if we also drop the failure check (which upstream phpbb code also doesn't do).
```php

```